relayer/relays/parachain/types.go

```diff
@@ -1,43 +1,131 @@
 package parachain
 
 import (
+	"fmt"
+	"math/bits"
+
 	"github.com/snowfork/go-substrate-rpc-client/v4/types"
 	"github.com/snowfork/snowbridge/relayer/chain/relaychain"
 	"github.com/snowfork/snowbridge/relayer/crypto/merkle"
 )
 
 // A Task is a bundle of items needed to submit commitments to Ethereum
 type Task struct {
-	ParaID                  uint32
-	BlockNumber             uint64
-	Header                  *types.Header
-	Commitments             map[ChannelID][]Commitment
-	BasicChannelBundleProof *MerkleProof
-	ProofInput              *ProofInput
-	ProofOutput             *ProofOutput
+	ParaID                        uint32
+	BlockNumber                   uint64
+	Header                        *types.Header
+	BasicChannelProofs            *[]MerkleProof
+	IncentivizedChannelCommitment *IncentivizedChannelCommitment
+	ProofInput                    *ProofInput
+	ProofOutput                   *ProofOutput
 }
 
 // A Commitment is data provably attested to by polkadot. The commitment hash
 // is contained in a parachain header. Polkadot validator nodes attest that the header
 // is genuine.
-type Commitment struct {
+type IncentivizedChannelCommitment struct {
 	Hash types.H256
-	Data interface{}
+	Data IncentivizedOutboundChannelMessageBundle
 }
 
-func NewCommitment(hash types.H256, data interface{}) Commitment {
-	return Commitment{
+func NewIncentivizedChannelCommitment(hash types.H256, data IncentivizedOutboundChannelMessageBundle) IncentivizedChannelCommitment {
+	return IncentivizedChannelCommitment{
 		Hash: hash,
 		Data: data,
 	}
 }
 
-type MerkleProof struct {
+type RawMerkleProof struct {
 	Root           types.H256
 	Proof          []types.H256
 	NumberOfLeaves uint64
 	LeafIndex      uint64
-	Leaf           []byte
+	// TODO: test that this decodes properly
+	Leaf []byte
+	// Leaf BasicOutboundChannelMessageBundle
+}
+
+type MerkleProof struct {
+	Root      types.H256
+	Proof     [][32]byte
+	HashSides []bool
+	Leaf      BasicOutboundChannelMessageBundle
+}
+
+func NewMerkleProof(rawProof RawMerkleProof, bundle BasicOutboundChannelMessageBundle) (MerkleProof, error) {
+	var proof MerkleProof
+
+	byteArrayProof := make([][32]byte, len(rawProof.Proof))
+	for i := 0; i < len(rawProof.Proof); i++ {
+		byteArrayProof[i] = ([32]byte)(rawProof.Proof[i])
+	}
+
+	hashSides, err := generateHashSides(rawProof)
+	if err != nil {
+		return proof, err
+	}
+
+	proof = MerkleProof{
+		Root:      rawProof.Root,
+		Proof:     byteArrayProof,
+		HashSides: hashSides,
+		Leaf:      bundle,
```

Comment:
I'm not sure how much we gain from this extra validation. The bundle in the event is the same data as in the proof's leaf. Both are emitted from the parachain, only difference is that the event is SCALE-encoded while the proof's leaf is ethabi-encoded.

Reply:
I'm leaning towards removing the leaf field (the message bundle) from the proof that the RPC returns.
That way, the only source of message bundles for relayers is the events. The events and proofs come from the parachain anyway, so we shouldn't need to validate them until we get to the inbound channel contract. If a bundle is paired with the wrong proof, the contract will reject it.

Reply:
I'm fine with that. I did intend for events querying to be the primary API for retrieving for bundles. :+1:

Though I would leave the leaf field in the RawMerkleProof RPC result. It may be useful for troubleshooting further down the line.

Reply:
I've left the leaf field in the raw proof, but removed it from `MerkleProof`. There's a `BundleProof` struct now that pairs a `MerkleProof` with its message bundle.